Fix verifying certs on windows

DennisDenuto opened the pull request:
Attempts to fix #22 

TODO:

- [x] document how to use custom ca certs with windows root store
- [x] figure out why windows still has a file handle on the bundle.tar file. TODO in test code
- [x] confirm ca certs work in an airgapped env
- [x] investigate if theres any performance cost due to using windows root ca store (I think it downloads ca certs using network)

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Add windows GH runner

- use ttl.sh as registry since registry:v2 doesn't have a windows image
- skip tests that cannot run on windows due to image (registry:2) used being available to linux

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
Fix git patch applying when using windows

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/imgpkg/cmd/push_test.go:
```diff
@@ -56,7 +57,7 @@ func TestMultiImgpkgDirError(t *testing.T) {
 		t.Fatalf("Expected validations to err, but did not")
 	}
 
-	if !strings.Contains(err.Error(), "This directory contains multiple bundle definitions. Only a single instance of .imgpkg/images.yml can be provided") {
+	if !strings.Contains(err.Error(), fmt.Sprintf("This directory contains multiple bundle definitions. Only a single instance of .imgpkg%simages.yml can be provided", string(os.PathSeparator))) {
```
im not sure how i feel about us flipping slashes in error messages. should we do this or just leave it as is?

DennisDenuto replied:
hmmm i'm not sure i fully understand here...

This is within a test asserting an error message though... so the implementation on windows is going to return a different error message when running this test in linux (the diff being the slash). So this aims to make the test make the right assertion based on the OS that is running said test.

cppforlife replied:
i see, we are generating that stuff on the fly.

DennisDenuto committed:
test refactor: fix test assertion error

- use human readable file permissions during test assertion failure message
- swap expected/actual test assertion values

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
test refactor: Stop using docker registry:v2 container to allow running on windows

- windows doesn't have a registry:v2 image
- run performance tests on windows

Authored-by: Dennis Leon <leonde@vmware.com>

DennisDenuto committed:
bug fix: Use the windows default ca store if custom ca certs are not provided by the user

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on pkg/imgpkg/registry/registry.go:
```diff
@@ -250,8 +251,17 @@ func (r Registry) FirstImageExists(digests []string) (string, error) {
-	if err != nil {
+	var pool *x509.CertPool
+
+	if runtime.GOOS != "windows" {
```
lets throw in a comment here why we are doing this.

DennisDenuto replied:
done

DennisDenuto committed:
Close chunked manifest.json file when reading from tarball

Authored-by: Dennis Leon <leonde@vmware.com>

cppforlife commented on test/helpers/assets.go:
```diff
@@ -94,6 +94,7 @@ func (a *Assets) CleanCreatedFolders() {
 	a.T.Helper()
 	for _, folder := range a.CreatedFolders {
 		err := os.RemoveAll(folder)
+
```
nit: remove newline?

DennisDenuto replied:
done

cppforlife commented on hack/build.sh:
```diff
@@ -2,6 +2,8 @@
 
 set -e -x -u
 
+FILE_EXT="${1-}"
```
instead of turning this into an arg wdyt about just using $IMGPKG_BINARY_EXT env var

DennisDenuto replied:
done

cppforlife commented on .github/workflows/gh-test.yml:
```diff
@@ -28,4 +28,36 @@ jobs:
+
+        export IMGPKG_E2E_IMAGE="ttl.sh/github-action-test-relocation-$GITHUB_RUN_ID"
+        export IMGPKG_E2E_RELOCATION_REPO="ttl.sh/github-action-imgpkg-test-$GITHUB_RUN_ID"
+        # deploy local registry and run tests
```
this comment doesnt seem to belong to anything

DennisDenuto replied:
true. removed.

cppforlife commented on test/e2e/gcloud_expired_token_test.go:
```diff
@@ -9,13 +9,17 @@ import (
+	if runtime.GOOS == "windows" {
+		t.Skip("Skipping test as docker image used requires linux")
+	}
 	env := helpers.BuildEnv(t)
```
nit: seems like in other places you gotta newline after this windows if

cppforlife commented on test/e2e/copy_from_image_test.go:
```diff
@@ -116,7 +113,7 @@ func TestCopyAnImageFromATarToARepoThatDoesNotContainNonDistributableLayersButTh
 		imgpkg.Run([]string{"copy", "-i", airgappedRepo, "--to-tar", tarFilePath})
 
-		stopRegistryForAirgapTesting(t, env)
+		stopRegistryForAirgapTesting(fakeRegistry)
```
should this function be replaced with defer fakeRegistry.CleanUp() after startRegistryForAirgapTesting()?

cppforlife replied:
oh i see what's going on you are clearing it before copy... nm

DennisDenuto committed:
test refactor: Use env var instead of bash arguments

- env var IMGPKG_BINARY_EXT instead of arg
- add new lines

Authored-by: Dennis Leon <leonde@vmware.com>
